spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/AsyncExecutor.scala

```diff
@@ -28,28 +30,43 @@ class AsyncExecutor[T, R](asyncAction: T => ListenableFuture[R], maxConcurrentTa
     pendingFutures.put(settable, true)
 
     val executionTimestamp = System.nanoTime()
-    val future = asyncAction(task)
-
-    Futures.addCallback(future, new FutureCallback[R] {
-      def release() {
-        semaphore.release()
-        pendingFutures.remove(settable)
-      }
-      def onSuccess(result: R) {
-        release()
-        settable.set(result)
-        successHandler.foreach(_(task, submissionTimestamp, executionTimestamp))
-      }
-      def onFailure(throwable: Throwable) {
-        logError("Failed to execute: " + task, throwable)
-        if (_successful) _successful = false
-        release()
-        settable.setException(throwable)
-        failureHandler.foreach(_(task, submissionTimestamp, executionTimestamp))
-      }
-    })
-
-    settable
+
+    def tryFuture(): SettableFuture[R] = {
+      val future = asyncAction(task)
+
+      Futures.addCallback(future, new FutureCallback[R] {
+        def release() {
+          semaphore.release()
+          pendingFutures.remove(settable)
+        }
+
+        def onSuccess(result: R) {
+          release()
+          settable.set(result)
+          successHandler.foreach(_ (task, submissionTimestamp, executionTimestamp))
+        }
+
+        def onFailure(throwable: Throwable) {
+          throwable match {
+            case nHAE: NoHostAvailableException if nHAE.getErrors.asScala.values.exists(_.isInstanceOf[BusyPoolException]) =>
+              logWarning("BusyPoolException ... Retrying")
+              tryFuture()
+
+            case otherException =>
+              logError("Failed to execute: " + task, throwable)
+              if (_successful) _successful = false
+              release()
+              settable.setException(throwable)
+              latestExeception = Some(throwable)
```

Comment:
Since `_successful` volatile write is conditional, we could make _successful not volatile and make latestException volatile. This may make Travis happy :)